pulsar-function-go/pf/instance.go

```diff
@@ -0,0 +1,289 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package pf
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/apache/pulsar/pulsar-client-go/pulsar"
+	"github.com/apache/pulsar/pulsar-function-go/log"
+	"github.com/apache/pulsar/pulsar-function-go/pb"
+)
+
+const PulsarServiceURL = "pulsar://localhost:6650"
+
+type GoInstance struct {
+	function  Function
+	context   *FunctionContext
+	producer  pulsar.Producer
+	consumers []pulsar.Consumer
+	client    pulsar.Client
+	msg       []pulsar.Message
+}
+
+func NewGoInstance() *GoInstance {
+	goInstance := &GoInstance{
+		context: NewFuncContext(),
+	}
+	return goInstance
+}
+
+type Handler struct {
+	handler Function
+}
+
+func (gi *GoInstance) handlerMsg() (output []byte, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	fc := NewFuncContext()
+	ctx = NewContext(ctx, fc)
+
+	if gi.msg == nil {
+		return nil, errors.New("no messages to be processed")
+	}
+	for _, msg := range gi.msg {
```

Comment:
so the logic here should only process one message.

```
ctx, cancel := context.WithCancel(context.Background())
defer cancel()
ctx = NewContext(ctx, gi.context)
msgInput := input.Payload()
return  gi.function.Process(ctx, msgInput)
```

Reply:
done